exit/test_exit.c

```diff
@@ -0,0 +1,528 @@
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		/* Try to kill process which exited */
+		errno = 0;
+		ret = kill(pid, SIGKILL);
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_EQUAL_INT(ESRCH, errno);
+	}
+}
+
+
+TEST(unistd_exit, close_streams)
+{
+	/*
+	 * Open pipe, when child exits one side of pipe should be closed,
+	 * so that call to write should fail.
+	 */
+	pid_t pid;
+	int pipefd[2];
+	int status;
+
+	TEST_ASSERT_EQUAL_INT(0, pipe(pipefd));
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int ret;
+
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		close(pipefd[0]); /* close read pipe end */
+		errno = 0;
+		ret = write(pipefd[1], TEST_STR, sizeof(TEST_STR));
+		TEST_ASSERT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_EQUAL_INT(EPIPE, errno);
+		close(pipefd[1]); /* close write pipe end */
+	}
+}
+
+
+TEST(unistd_exit, orphaned_child)
+{
+	/* Test if parent _exit affect child process */
+	pid_t pid;
+	int pipefd[2];
+
+	/* Pipe needed for communication between grandparent and grandchild (can't use asserts in child) */
+	TEST_ASSERT_EQUAL_INT(0, pipe(pipefd));
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	if (pid == 0) {
+		pid = fork();
+		/* grandchild */
+		if (pid == 0) {
+			/* Grandchild doesn't need pipe */
+			close(pipefd[0]);
+			close(pipefd[1]);
+
+			/* Sleep for a while to ensure that parent already exited */
+			sleep(1);
+			creat(PATH3, DEFFILEMODE);
+			_exit(EXIT_SUCCESS);
+		}
+		/* parent */
+		else {
+			/* Write fork return value from parent to grandparent */
+			close(pipefd[0]);
+			write(pipefd[1], &pid, sizeof(pid_t));
+			close(pipefd[1]);
+			/* parent exits right away */
+			_exit(EXIT_SUCCESS);
+		}
+	}
+	/* grandparent */
+	else {
+		int status;
+
+		close(pipefd[1]);
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		read(pipefd[0], &pid, sizeof(pid_t));
+		TEST_ASSERT_NOT_EQUAL_INT(-1, pid);
+		close(pipefd[0]);
+
+		/* Check if file exists, if so child sleep was interrupted */
+		TEST_ASSERT_EQUAL_INT(-1, access(PATH3, F_OK));
+
+		sleep(2);
+		/* At this moment if child wasn't affected file should exist */
+		TEST_ASSERT_EQUAL_INT(0, access(PATH3, F_OK));
+
+		remove(PATH3);
+	}
+}
+
+
+TEST(unistd_exit, new_parent_id)
+{
+	/* Test that child acquire new parent ID */
+	pid_t pid;
+	int pipefd[2];
+
+	/* Pipe needed for communication between grandparent and grandchild (can't use asserts in child) */
+	TEST_ASSERT_EQUAL_INT(0, pipe(pipefd));
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	if (pid == 0) {
+		pid = fork();
+		/* grandchild */
+		if (pid == 0) {
+			pid_t old_ppid, new_ppid;
+
+			close(pipefd[0]);
+			old_ppid = getppid();
+			/* Sleep until parents exits */
+			sleep(2);
```

Comment:
I think we can use much lower sleep values, you can consider using usleep - it may make this test much faster.

Reply:
I got rid of sleeps